File: SwiftUI/EdgeStudio/Models/LogEntry.swift
@preconcurrency import DittoSwift
import Foundation

/// A single log entry from any source (Ditto SDK, app logs, or imported files).
struct LogEntry: Identifiable {
    let id: UUID
    let timestamp: Date
    let level: DittoLogLevel
    let message: String
    let component: LogComponent
    let source: LogEntrySource
    /// Original raw text line for copy/export
    let rawLine: String

    init(
        id: UUID = UUID(),
        timestamp: Date,
        level: DittoLogLevel,
        message: String,
        component: LogComponent = .other,
        source: LogEntrySource,
        rawLine: String
    ) {
        self.id = id
        self.timestamp = timestamp
        self.level = level
        self.message = message
        self.component = component
        self.source = source
        self.rawLine = rawLine
    }
}

extension LogEntry {
    /// Returns true if the entry's timestamp falls within [start, end] (inclusive on both ends).
    static func isWithinDateRange(_ entry: LogEntry, start: Date, end: Date) -> Bool {
        entry.timestamp >= start && entry.timestamp <= end
    }
}

/// Describes where a log entry originated.
enum LogEntrySource: Equatable {
    case dittoSDK
    case application
    case imported(label: String)
}

/// Log component, mapped from the Ditto SDK `target` field or inferred heuristically.
enum LogComponent: String, CaseIterable {
    case all = "All"
    case sync = "Sync"
    case store = "Store"
    case query = "Query"
    case observer = "Observer"
    case transport = "Transport"
    case auth = "Auth"
    case other = "Other"

    /// Maps a Ditto SDK `target` field (e.g. "ditto::sync") to a component.
    static func from(target: String) -> LogComponent {
        let lower = target.lowercased()
        if lower.contains("sync") { return .sync }
        if lower.contains("replication") { return .sync }
        if lower.contains("subscription") { return .sync }
        if lower.contains("store") { return .store }
        if lower.contains("service=blob") { return .store }
        if lower.contains("query") { return .query }
        if lower.contains("sqlparser") || lower.contains("sql_parser") { return .query }
        if lower.contains("observer") { return .observer }
        if lower.contains("transport") { return .transport }
        if lower.contains("discovery") { return .transport }
        if lower.contains("presence") { return .transport }
        if lower.contains("multihop") { return .transport }
        if lower.contains("network") { return .transport }
        if lower.contains("ble") { return .transport }
        if lower.contains("tcp") { return .transport }
        if lower.contains("awdl") { return .transport }
        if lower.contains("virtual_connection") { return .transport }
        if lower.contains("router") { return .transport }
        if lower.contains("auth") { return .auth }
        return .other
    }

    /// Heuristic component detection from plain-text callback messages.
    static func heuristic(from message: String) -> LogComponent {
        let lower = message.lowercased()
        if lower.contains("sync") { return .sync }
        if lower.contains("replication") { return .sync }
        if lower.contains("subscription") { return .sync }
        if lower.contains("store") || lower.contains("insert") || lower.contains("document") { return .store }
        if lower.contains("service=blob") { return .store }
        // Transport-first: well-known SDK operation names that must not be hijacked by a
        // "query" substring appearing later in the long message body.
        if lower.hasPrefix("add_ble_transport") ||
            lower.hasPrefix("start_tcp_server") ||
            lower.hasPrefix("add_awdl_transport") ||
            lower.hasPrefix("add_wifi_transport") { return .transport }
        // Additional missing transport keywords
        if lower.contains("tcp") { return .transport }
        if lower.contains("awdl") { return .transport }
        if lower.contains("query") || lower.contains("select") { return .query }
        if lower.hasPrefix("parsing sql") || lower.contains("sql parser") { return .query }
        if lower.contains("observer") { return .observer }
        if lower.contains("transport") || lower.contains("bluetooth") || lower.contains("wifi") { return .transport }
        if lower.contains("discovery") || lower.contains("mdns") { return .transport }
        if lower.contains("presence") || lower.contains("multihop") { return .transport }
        if lower.contains("ble_") || lower.contains(" ble") { return .transport }
        if lower.contains("virtual_connection") { return .transport }
        if lower.contains("router_") { return .transport }
        if lower.contains("auth") || lower.contains("token") { return .auth }
        return .other
    }
}

extension DittoLogLevel {
    /// Human-readable label for display in the UI.
    var displayName: String {
        switch self {
        case .error: return "Error"
        case .warning: return "Warning"
        case .info: return "Info"
        case .debug: return "Debug"
        case .verbose: return "Verbose"
        @unknown default: return "Unknown"
        }
    }

    /// Short label for compact display.
    var shortName: String {
        switch self {
        case .error: return "ERR"
        case .warning: return "WARN"
        case .info: return "INFO"
        case .debug: return "DBG"
        case .verbose: return "VERB"
        @unknown default: return "?"
        }
    }

    /// Initializes from the string stored in `DittoConfigForDatabase.logLevel`.
    static func from(string: String) -> DittoLogLevel {
        switch string {
        case "error": return .error
        case "warning": return .warning
        case "debug": return .debug
        case "verbose": return .verbose
        default: return .info
        }
    }

    /// Returns a string suitable for storage in `DittoConfigForDatabase.logLevel`.
    var storageString: String {
        switch self {
        case .error: return "error"
        case .warning: return "warning"
        case .info: return "info"
        case .debug: return "debug"
        case .verbose: return "verbose"
        @unknown default: return "info"
        }
    }
}

File: android/app/src/main/java/com/costoda/dittoedgestudio/domain/model/LogEntry.kt
package com.costoda.dittoedgestudio.domain.model

import com.ditto.kotlin.DittoLogLevel
import java.util.Date
import java.util.UUID

data class LogEntry(
    val id: UUID = UUID.randomUUID(),
    val timestamp: Date,
    val level: DittoLogLevel,
    val message: String,
    val component: LogComponent,
    val source: LogEntrySource,
    val rawLine: String,
)

enum class LogComponent(val displayName: String) {
    ALL("All"),
    SYNC("Sync"),
    STORE("Store"),
    QUERY("Query"),
    OBSERVER("Observer"),
    TRANSPORT("Transport"),
    AUTH("Auth"),
    OTHER("Other");

    companion object {
        /**
         * Maps a Ditto SDK `target` field (e.g. `ditto_discovery_mdns`) to a component.
         *
         * **The substring list and its order are normative** — they mirror
         * `SwiftUI/EdgeStudio/Models/LogEntry.swift` `LogComponent.from(target:)`
         * one-for-one. Order is load-bearing: a target such as
         * `ditto_sync::query_planner` must resolve to [SYNC] because `sync` is
         * tested first, and `service=blob` must resolve to [STORE] before any
         * later transport token can claim it. Reordering these branches silently
         * reclassifies records.
         *
         * Before this list matched the Swift one, 866 of 746 282 real captured
         * records classified differently — every one of them Transport on
         * SwiftUI and [OTHER] here (`ditto_discovery_mdns`,
         * `ditto_discovery_multicast[::interface]`,
         * `ditto_presence::multihop::manager`).
         */
        fun from(target: String): LogComponent {
            val lower = target.lowercase()
            return when {
                "sync" in lower -> SYNC
                "replication" in lower -> SYNC
                "subscription" in lower -> SYNC
                "store" in lower -> STORE
                "service=blob" in lower -> STORE
                "query" in lower -> QUERY
                "sqlparser" in lower || "sql_parser" in lower -> QUERY
                "observer" in lower -> OBSERVER
                "transport" in lower -> TRANSPORT
                "discovery" in lower -> TRANSPORT
                "presence" in lower -> TRANSPORT
                "multihop" in lower -> TRANSPORT
                "network" in lower -> TRANSPORT
                "ble" in lower -> TRANSPORT
                "tcp" in lower -> TRANSPORT
                "awdl" in lower -> TRANSPORT
                "virtual_connection" in lower -> TRANSPORT
                "router" in lower -> TRANSPORT
                "auth" in lower -> AUTH
                else -> OTHER
            }
        }

        /**
         * Heuristic component detection from the flattened live-callback message
         * body, which carries no `target` field.
         *
         * Mirrors `LogComponent.heuristic(from:)` in the SwiftUI reference,
         * including its order. The transport branches ahead of `query` are
         * deliberate: SDK operation names such as `start_tcp_server` and
         * `add_awdl_transport` appear in long message bodies that also happen to
         * contain the word "query", and without the transport-first probes those
         * lines would be filed under [QUERY].
         */
        fun heuristic(message: String): LogComponent {
            val lower = message.lowercase()
            return when {
                "sync" in lower -> SYNC
                "replication" in lower -> SYNC
                "subscription" in lower -> SYNC
                "store" in lower || "insert" in lower || "document" in lower -> STORE
                "service=blob" in lower -> STORE
                // Transport-first: well-known SDK operation names that must not be
                // hijacked by a "query" substring appearing later in the body.
                lower.startsWith("add_ble_transport") ||
                    lower.startsWith("start_tcp_server") ||
                    lower.startsWith("add_awdl_transport") ||
                    lower.startsWith("add_wifi_transport") -> TRANSPORT
                "tcp" in lower -> TRANSPORT
                "awdl" in lower -> TRANSPORT
                "query" in lower || "select" in lower -> QUERY
                lower.startsWith("parsing sql") || "sql parser" in lower -> QUERY
                "observer" in lower -> OBSERVER
                "transport" in lower || "bluetooth" in lower || "wifi" in lower -> TRANSPORT
                "discovery" in lower || "mdns" in lower -> TRANSPORT
                "presence" in lower || "multihop" in lower -> TRANSPORT
                "ble_" in lower || " ble" in lower -> TRANSPORT
                "virtual_connection" in lower -> TRANSPORT
                "router_" in lower -> TRANSPORT
                "auth" in lower || "token" in lower -> AUTH
                else -> OTHER
            }
        }
    }
}

sealed class LogEntrySource {
    object DittoSDK : LogEntrySource()
    object Application : LogEntrySource()

    /** Transport-condition events from `Ditto.transportCondition` (SwiftUI parity). */
    object TransportConditions : LogEntrySource()

    /** Connection-request events from `presence.connectionRequestHandler` (SwiftUI parity). */
    object ConnectionRequests : LogEntrySource()
}

val DittoLogLevel.displayName: String
    get() = when (this) {
        DittoLogLevel.Error -> "Error"
        DittoLogLevel.Warning -> "Warning"
        DittoLogLevel.Info -> "Info"
        DittoLogLevel.Debug -> "Debug"
        DittoLogLevel.Verbose -> "Verbose"
    }

val DittoLogLevel.shortName: String
    get() = when (this) {
        DittoLogLevel.Error -> "ERR"
        DittoLogLevel.Warning -> "WARN"
        DittoLogLevel.Info -> "INFO"
        DittoLogLevel.Debug -> "DBG"
        DittoLogLevel.Verbose -> "VERB"
    }
